core/src/main/java/org/apache/iceberg/io/PartitioningWriter.java

```diff
@@ -46,7 +46,7 @@
    * @param spec a partition spec
    * @param partition a partition or null if the spec is unpartitioned
    */
-  void write(T row, PartitionSpec spec, StructLike partition);
+  PathOffset write(T row, PartitionSpec spec, StructLike partition);
```

Comment:
As the `PartitioningWriter` will have multiple partition writers in its internal implementation,  so we can not just introduce the two methods in this interface: 

```java
  /**
   * Returns the file path that are currently opened.
   *
   * @return the current file path.
   */
  CharSequence location();

  /**
   * Returns the row offset that are currently writing, starting from 0.
   *
   * @return the current row offset.
   */
  long rowOffset();
```

Because those `location()` and `rowOffset()` are binded to specific writers. In theory, different writers should have different `location()` and `rowOffset()`.  So here we add a return value `PathOffset` here to get the latest wrote path & offset for the wrote row.